pandas/core/indexes/multi.py

```diff
@@ -3601,6 +3601,8 @@ def intersection(self, other, sort=False):
         other, result_names = self._convert_can_do_setop(other)
 
         if self.equals(other):
+            if self.has_duplicates:
+                return self.unique().rename(result_names)
```

Comment:
@phofl it looks like we do something slightly different in many of the `FooIndex.intersection` methods in the `self.equals(other)` case:

```
# Index (and IntervalIndex pending #38190)
        if self.equals(other) and not self.has_duplicates:
            return self._get_reconciled_name_object(other)

# datetimelike
        if self.equals(other):
            return self._get_reconciled_name_object(other)

# MultiIndex
        if self.equals(other):
            if self.has_duplicates:
                return self.unique().rename(result_names)
            return self._get_reconciled_name_object(other)

# PeriodIndex
        if self.equals(other):
            return self._get_reconciled_name_object(other)

# RangeIndex
        if self.equals(other):
            return self._get_reconciled_name_object(other)
```

The RangeIndex one is equivalent to the Index/Intervalindex bc it never has duplicates (can add that check to make the code match exactly).  Can the others be made to have identical logic?

Reply:
Is there still an open case where I can be of help?

Reply:
I think we're all set for now, will take another look after all the current Index.intersection PRs go through.  thanks

Reply:
Great

Reply:
It looks like several of them are still either

```
        if self.equals(other):
            return self._get_reconciled_name_object(other)
```

which i think is wrong if it has duplicates, or

```
        if self.equals(other) and not self.has_duplicates:
            return self._get_reconciled_name_object(other)
```

which isnt handling the has-duplicates case like the others.  can these all be identical?

Also if you really get on a roll, `DatetimeTimedeltaMixin._intersection` has cases for `len(self) == 0` and `len(other)==0` that would be nice to standardize+test.  RangeIndex._intersection has a similar check.

Reply:
Will look through them and try to standardize them as much as possible